shared/src/ledger/native_vp/masp.rs

```diff
@@ -396,29 +367,91 @@ where
-                    denom,
-                );
-
-                // Non-masp sources add to transparent tx pool
-                transparent_tx_pool += transp_amt;
-            }
-
             // No shielded spends nor shielded converts are allowed
             if shielded_tx.sapling_bundle().is_some_and(|bundle| {
                 !(bundle.shielded_spends.is_empty()
                     && bundle.shielded_converts.is_empty())
             }) {
                 return Ok(false);
             }
+
+            let transp_bundle =
+                shielded_tx.transparent_bundle().ok_or_err_msg(
+                    "Expected transparent outputs in shielding transaction",
+                )?;
+            let mut total_in_values = token::Amount::zero();
+            let source_enc = transfer.source.serialize_to_vec();
+            let hash =
+                ripemd::Ripemd160::digest(sha2::Sha256::digest(&source_enc));
+
+            // To help recognize asset types not in the conversion tree
+            let unepoched_tokens = unepoched_tokens(&transfer.token)?;
+            // Handle transparent input
+            // The following boundary conditions must be satisfied
+            // 1. Total of transparent input values equals containing transfer
+            // amount 2. Asset type must be properly derived
+            // 3. Public key must be the hash of the source
+            for vin in &transp_bundle.vin {
+                // Non-masp sources add to the transparent tx pool
+                transparent_tx_pool += I128Sum::from_nonnegative(
+                    vin.asset_type,
+                    vin.value as i128,
+                )
+                .ok()
+                .ok_or_err_msg("invalid value or asset type for amount")?;
+
+                // Satisfies 3.
+                if <[u8; 20]>::from(hash) != vin.address.0 {
+                    tracing::debug!(
+                        "the public key of the output account does not match \
+                         the transfer target"
+                    );
+                    return Ok(false);
+                }
+                match conversion_state.assets.get(&vin.asset_type) {
+                    // Satisfies 2.
+                    Some(((address, denom), asset_epoch, _, _))
+                        if *address == transfer.token
+                            && *asset_epoch <= epoch =>
+                    {
+                        total_in_values += token::Amount::from_masp_denominated(
+                            vin.value, *denom,
+                        );
+                    }
+                    // Maybe the asset type has no attached epoch
+                    None if unepoched_tokens.contains_key(&vin.asset_type) => {
+                        let (token, denom) = &unepoched_tokens[&vin.asset_type];
+                        // Determine what the asset type would be if it were
+                        // epoched
+                        let epoched_asset_type =
```

Comment:
Is there no danger that this note is in the conversion with an epoch other than the checked against here? Or is that considered fine, since I guess we upconvert in the else branch?

Reply:
I believe once an asset is in the `ConversionState`, the function `update_allowed_conversions` will update the conversions at every new epoch, so probably we are guaranteed that either the asset is in the conversion state and we have the conversion with the latest epoch (the one we pass here to `encode_asset_type`) or we don't have it at all?

Reply:
Yes, this is exactly how I was reasoning. Though this code could be a little bit more robust...